composer/2022_airflow_summit/data_analytics_process_expansion.py

```diff
@@ -0,0 +1,221 @@
+from pyspark.sql.types import FloatType, IntegerType, StructType
+
+# BQ_DESTINATION_DATASET_NAME = "expansion_project"
+# BQ_DESTINATION_TABLE_NAME = "ghcnd_stations_joined"
+# BQ_NORMALIZED_TABLE_NAME = "ghcnd_stations_normalized"
+# BQ_PRCP_MEAN_TABLE_NAME = "ghcnd_stations_prcp_mean"
+# BQ_SNOW_MEAN_TABLE_NAME = "ghcnd_stations_snow_mean"
+# BQ_PHX_PRCP_TABLE_NAME = "phx_annual_prcp"
+# BQ_PHX_SNOW_TABLE_NAME = "phx_annual_snow"
+
+if __name__ == "__main__":
+    BUCKET_NAME = sys.argv[1]
+    READ_TABLE = sys.argv[2]
+    DF_WRITE_TABLE = sys.argv[3]
+    PRCP_MEAN_WRITE_TABLE = sys.argv[4]
+    SNOW_MEAN_WRITE_TABLE = sys.argv[5]
+    PHX_PRCP_WRITE_TABLE = sys.argv[6]
+    PHX_SNOW_WRITE_TABLE = sys.argv[7]
+
+    # BUCKET_NAME = "workshop_example_bucket"
+    # READ_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_DESTINATION_TABLE_NAME}"
+    # DF_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_NORMALIZED_TABLE_NAME}"
+    # PRCP_MEAN_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_PRCP_MEAN_TABLE_NAME}"
+    # SNOW_MEAN_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_SNOW_MEAN_TABLE_NAME}"
+    # PHX_PRCP_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_PHX_PRCP_TABLE_NAME}"
+    # PHX_SNOW_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_PHX_SNOW_TABLE_NAME}"
+
+    # Create a SparkSession, viewable via the Spark UI
+    spark = SparkSession.builder.appName("data_processing").getOrCreate()
+    # Load data into dataframe if READ_TABLE exists
+    try:
+        df = spark.read.format("bigquery").load(READ_TABLE)
+    except Py4JJavaError as e:
+        raise Exception(f"Error reading {READ_TABLE}") from e
+
+    # filter out non-west states of the US
+    df = df.where((df.STATE == 'WA') | (df.STATE == 'OR') | (df.STATE == 'ID')
+                  | (df.STATE == 'MT') | (df.STATE == 'WY') | (df.STATE == 'CO')
+                  | (df.STATE == 'NM') | (df.STATE == 'AZ') | (df.STATE == 'UT')
+                  | (df.STATE == 'NV') | (df.STATE == 'CA'))
+    df.show(n=10)
+    print("After state filtering, # of rows remaining is:", df.count())
+
+    # filter out rows whose element value does not equal to PRCP or SNOW
+    df = df.where((df.ELEMENT == 'PRCP') | (df.ELEMENT == 'SNOW'))
+    df.show(n=10)
+    print("After elemet filtering, # of rows remaining is:", df.count())
+
+    # convert precipitation and snowfall from "tenths of a mm" to "mm"
+    @udf(returnType=FloatType())
+    def converter(val) -> float:
+        return val / 10
+    df = df.withColumn("VALUE", converter(df.VALUE))
+    df.show()
+
+    # return the year of each date
+    @udf(returnType=IntegerType())
+    def date_to_year(val) -> int:
+        return val.year
+
+    # extract the year of each date and rename it as YEAR
+    # This will allow us to merge the data based on the years they are created instead of date
+    df = df.withColumn("DATE", date_to_year(df.DATE)).withColumnRenamed('DATE', 'YEAR')
+
+    # each year's arithmetic mean of pricipation
```

Comment:
pricipation -> precipitation (the spelling is tricky, I had to look it up myself)

Reply:
oh lol i didn't even notice this because my brain read it correctly. this word is annoying to spell

Reply:
Done!